File: packages/pin/ios/error/WepinError.swift
import Foundation
import WepinCommon

internal class WepinErrorMapper {
  // WepinError(Swift)를 React Native 에러 코드와 메시지로 변환하는 메서드
  static func getCodeAndMessage(_ error: Error) -> (String, String) {
    // WepinError 타입으로 캐스팅 가능한 경우
    if let wepinError = error as? WepinError {
      let loginError = mapToLoginError(wepinError)
      return (loginError.code.rawValue, loginError.message)
    }
    
    // 그 외 일반 에러
    return (WepinReactErrorCode.UnknownError.code, error.localizedDescription)
  }
  
  // WepinError를 React Native에서 사용할 수 있는 에러 형식으로 변환
  private static func mapToLoginError(_ error: WepinError) -> WepinReactError {
    let errorCode: WepinReactErrorCode
    
    switch error {
    case .alreadyInitialized:
      errorCode = .AlreadyInitialized
    case .alreadyLogout:
      errorCode = .AlreadyLogout
    case .apiRequestError:
      errorCode = .ApiRequestError
    case .invalidParameter:
      errorCode = .InvalidParameters
    case .notInitialized:
      errorCode = .NotInitialized
    case .invalidAppKey:
      errorCode = .InvalidAppKey
    case .invalidLoginProvider:
      errorCode = .InvalidLoginProvider
    case .invalidToken:
      errorCode = .InvalidToken
    case .invalidLoginSession, .invalidLoginSessionSimple:
      errorCode = .InvalidLoginSession
    case .userCanceled:
      errorCode = .UserCancelled
    case .notConnectedInternet:
      errorCode = .NotConnectedInternet
    case .loginFailed:
      errorCode = .FailedLogin
    case .requiredEmailVerified:
      errorCode = .RequiredEmailVerified
    case .incorrectEmailForm:
      errorCode = .IncorrectEmailForm
    case .incorrectPasswordForm:
      errorCode = .IncorrectPasswordForm
    case .networkNotInitialized:
      errorCode = .NotInitializedNetwork
    case .requiredSignupEmail:
      errorCode = .RequiredSignupEmail
    case .failedEmailVerification:
      errorCode = .FailedEmailVerified
    case .failedPasswordStateSetting:
      errorCode = .FailedPasswordStateSetting
    case .failedPasswordSetting:
      errorCode = .FailedPasswordSetting
    case .existedEmail:
      errorCode = .ExistedEmail
    case .deprecated:
      errorCode = .Deprecated
    case .invalidViewController:
      errorCode = .NotActivity
    case .invalidEmailDomain:
      errorCode = .InvalidEmailDomain
    case .failedSendEmail:
      errorCode = .FailedSendEmail
    default:
      errorCode = .UnknownError
    }
    
    return WepinReactError(code: errorCode, message: error.errorDescription ?? "Unknown error")
  }
  
  // React Native 에러 처리를 위해 WepinError를 RCTPromiseRejectBlock으로 변환
  static func rejectWithError(_ error: Error, reject: RCTPromiseRejectBlock) {
    let (code, message) = getCodeAndMessage(error)
    reject(code, message, error as NSError)
  }
}

// React Native에서 사용할 에러 코드
enum WepinReactErrorCode: String {
  case ApiRequestError = "ApiRequestError"
  case InvalidParameters = "InvalidParameters"
  case NotInitialized = "NotInitialized"
  case InvalidAppKey = "InvalidAppKey"
  case InvalidLoginProvider = "InvalidLoginProvider"
  case InvalidToken = "InvalidToken"
  case InvalidLoginSession = "InvalidLoginSession"
  case UserCancelled = "UserCancelled"
  case NotConnectedInternet = "NotConnectedInternet"
  case FailedLogin = "FailedLogin"
  case AlreadyLogout = "AlreadyLogout"
  case AlreadyInitialized = "AlreadyInitialized"
  case InvalidEmailDomain = "InvalidEmailDomain"
  case FailedSendEmail = "FailedSendEmail"
  case RequiredEmailVerified = "RequiredEmailVerified"
  case IncorrectEmailForm = "IncorrectEmailForm"
  case IncorrectPasswordForm = "IncorrectPasswordForm"
  case NotInitializedNetwork = "NotInitializedNetwork"
  case RequiredSignupEmail = "RequiredSignupEmail"
  case FailedEmailVerified = "FailedEmailVerified"
  case FailedPasswordStateSetting = "FailedPasswordStateSetting"
  case FailedPasswordSetting = "FailedPasswordSetting"
  case ExistedEmail = "ExistedEmail"
  case NotActivity = "NotActivity"
  case Deprecated = "Deprecated"
  case UnknownError = "UnknownError"
  
  var code: String {
    return self.rawValue
  }
}

// 변환된 에러 모델
struct WepinReactError {
  let code: WepinReactErrorCode
  let message: String
}

File: packages/pin/android/src/main/java/com/wepin/pinrn/error/WepinError.kt
package com.wepin.pinrn.error

import com.wepin.android.commonlib.error.WepinError

object WepinErrorMapper {
  private fun mapToLoginError(error: WepinError): WepinReactError {
    val errorCode = when (error.getErrorCode()) {
      WepinError.API_REQUEST_ERROR.code -> WepinReactErrorCode.ApiRequestError
      WepinError.INVALID_PARAMETER.code -> WepinReactErrorCode.InvalidParameters
      WepinError.NOT_INITIALIZED_ERROR.code -> WepinReactErrorCode.NotInitialized
      WepinError.INVALID_APP_KEY.code -> WepinReactErrorCode.InvalidAppKey
      WepinError.INVALID_LOGIN_PROVIDER.code -> WepinReactErrorCode.InvalidLoginProvider
      WepinError.INVALID_TOKEN.code -> WepinReactErrorCode.InvalidToken
      WepinError.INVALID_LOGIN_SESSION.code -> WepinReactErrorCode.InvalidLoginSession
      WepinError.USER_CANCELED.code -> WepinReactErrorCode.UserCancelled
      WepinError.NOT_CONNECTED_INTERNET.code -> WepinReactErrorCode.NotConnectedInternet
      WepinError.FAILED_LOGIN.code -> WepinReactErrorCode.FailedLogin
      WepinError.ALREADY_LOGOUT.code -> WepinReactErrorCode.AlreadyLogout
      WepinError.ALREADY_INITIALIZED_ERROR.code -> WepinReactErrorCode.AlreadyInitialized
      WepinError.INVALID_EMAIL_DOMAIN.code -> WepinReactErrorCode.InvalidEmailDomain
      WepinError.FAILED_SEND_EMAIL.code -> WepinReactErrorCode.FailedSendEmail
      WepinError.REQUIRED_EMAIL_VERIFIED.code -> WepinReactErrorCode.RequiredEmailVerified
      WepinError.INCORRECT_EMAIL_FORM.code -> WepinReactErrorCode.IncorrectEmailForm
      WepinError.INCORRECT_PASSWORD_FORM.code -> WepinReactErrorCode.IncorrectPasswordForm
      WepinError.NOT_INITIALIZED_NETWORK.code -> WepinReactErrorCode.NotInitializedNetwork
      WepinError.REQUIRED_SIGNUP_EMAIL.code -> WepinReactErrorCode.RequiredSignupEmail
      WepinError.FAILED_EMAIL_VERIFICATION.code -> WepinReactErrorCode.FailedEmailVerified
      WepinError.FAILED_PASSWORD_STATE_SETTING.code -> WepinReactErrorCode.FailedPasswordStateSetting
      WepinError.FAILED_PASSWORD_SETTING.code -> WepinReactErrorCode.FailedPasswordSetting
      WepinError.EXISTED_EMAIL.code -> WepinReactErrorCode.ExistedEmail
      WepinError.NOT_ACTIVITY.code -> WepinReactErrorCode.NotActivity
      else -> WepinReactErrorCode.UnknownError
    }

    return WepinReactError(
      code = errorCode,
      message = error.errorMessage ?: error.message
    )
  }

  fun getCodeAndMessage(error: WepinError): Pair<String, String?> {
    val mapped = mapToLoginError(error)
    return Pair(mapped.code.code, mapped.message)
  }
}

data class WepinReactError(
  val code: WepinReactErrorCode,
  val message: String? = null
) {
  override fun toString(): String {
    return "WepinLoginError(code=${code.code}, message=$message)"
  }
}

enum class WepinReactErrorCode(val code: String) {
  ApiRequestError("ApiRequestError"),
  InvalidParameters("InvalidParameters"),
  NotInitialized("NotInitialized"),
  InvalidAppKey("InvalidAppKey"),
  InvalidLoginProvider("InvalidLoginProvider"),
  InvalidToken("InvalidToken"),
  InvalidLoginSession("InvalidLoginSession"),
  UserCancelled("UserCancelled"),
  UnknownError("UnknownError"),
  NotConnectedInternet("NotConnectedInternet"),
  FailedLogin("FailedLogin"),
  AlreadyLogout("AlreadyLogout"),
  AlreadyInitialized("AlreadyInitialized"),
  InvalidEmailDomain("InvalidEmailDomain"),
  FailedSendEmail("FailedSendEmail"),
  RequiredEmailVerified("RequiredEmailVerified"),
  IncorrectEmailForm("IncorrectEmailForm"),
  IncorrectPasswordForm("IncorrectPasswordForm"),
  NotInitializedNetwork("NotInitializedNetwork"),
  RequiredSignupEmail("RequiredSignupEmail"),
  FailedEmailVerified("FailedEmailVerified"),
  FailedPasswordStateSetting("FailedPasswordStateSetting"),
  FailedPasswordSetting("FailedPasswordSetting"),
  ExistedEmail("ExistedEmail"),
  NotActivity("NotActivity")
}
